cli/factory.rs

```diff
@@ -0,0 +1,669 @@
+use crate::module_loader::CjsResolutionStore;
+use crate::module_loader::CliModuleLoaderFactory;
+use crate::module_loader::ModuleLoadPreparer;
+use crate::module_loader::NpmModuleLoader;
+use crate::node::CliCjsEsmCodeAnalyzer;
+use crate::node::CliNodeCodeTranslator;
+use crate::npm::create_npm_fs_resolver;
+use crate::npm::CliNpmRegistryApi;
+use crate::npm::CliNpmResolver;
+use crate::npm::NpmCache;
+use crate::npm::NpmResolution;
+use crate::npm::PackageJsonDepsInstaller;
+use crate::resolver::CliGraphResolver;
+use crate::tools::check::TypeChecker;
+use crate::util::progress_bar::ProgressBar;
+use crate::util::progress_bar::ProgressBarStyle;
+use crate::watcher::FileWatcher;
+use crate::watcher::FileWatcherReporter;
+use crate::worker::CliMainWorkerFactory;
+use crate::worker::CliMainWorkerOptions;
+use crate::worker::HasNodeSpecifierChecker;
+
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+
+use deno_runtime::deno_node;
+use deno_runtime::deno_node::analyze::NodeCodeTranslator;
+use deno_runtime::deno_node::NodeResolver;
+use deno_runtime::deno_tls::rustls::RootCertStore;
+use deno_runtime::deno_web::BlobStore;
+use deno_runtime::inspector_server::InspectorServer;
+use deno_semver::npm::NpmPackageReqReference;
+use import_map::ImportMap;
+use log::warn;
+use std::cell::RefCell;
+use std::future::Future;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+pub struct CliFactoryBuilder {
+  maybe_sender: Option<tokio::sync::mpsc::UnboundedSender<Vec<PathBuf>>>,
+}
+
+impl CliFactoryBuilder {
+  pub fn new() -> Self {
+    Self { maybe_sender: None }
+  }
+
+  pub fn with_watcher(
+    mut self,
+    sender: tokio::sync::mpsc::UnboundedSender<Vec<PathBuf>>,
+  ) -> Self {
+    self.maybe_sender = Some(sender);
+    self
+  }
+
+  pub async fn build_from_flags(
+    self,
+    flags: Flags,
+  ) -> Result<CliFactory, AnyError> {
+    Ok(self.build_from_cli_options(Arc::new(CliOptions::from_flags(flags)?)))
+  }
+
+  pub fn build_from_cli_options(self, options: Arc<CliOptions>) -> CliFactory {
+    CliFactory {
+      maybe_sender: RefCell::new(self.maybe_sender),
+      options,
+      services: Default::default(),
+    }
+  }
+}
+
+struct Deferred<T>(once_cell::unsync::OnceCell<T>);
+
+impl<T> Default for Deferred<T> {
+  fn default() -> Self {
+    Self(once_cell::unsync::OnceCell::default())
+  }
+}
+
+impl<T> Deferred<T> {
+  pub fn get_or_try_init(
+    &self,
+    create: impl FnOnce() -> Result<T, AnyError>,
+  ) -> Result<&T, AnyError> {
+    self.0.get_or_try_init(create)
+  }
+
+  pub fn get_or_init(&self, create: impl FnOnce() -> T) -> &T {
+    self.0.get_or_init(create)
+  }
+
+  pub async fn get_or_try_init_async(
+    &self,
+    create: impl Future<Output = Result<T, AnyError>>,
+  ) -> Result<&T, AnyError> {
+    if self.0.get().is_none() {
+      // todo(dsherret): it would be more ideal if this enforced a
+      // single executor and then we could make some initialization
+      // concurrent
```

Comment:
I'm not sure I understand this comment. Could you elaborate?

Reply:
Right now if someone were to create something async in parallel (ex. `std::futures::join_all(&[factory.something(), factory.something()])`) and it would create the result twice, but throw away one result. It would be better to just have it create the result once (so use something like an AsyncLazy here). That said, nothing is creating stuff concurrently at the moment so it doesn't matter, but it might be a future optimization if we see a place that would make sense to take advantage of concurrency.

Reply:
Makes sense, thanks for the explanation. In other words it's a footgun for ourselves. Proceed with care.